Add Light Metal capture/replay initial changes to tt-metal for some workloads

kmabeeTT opened the pull request:
### Ticket

> [Feature Request] Add Light Metal capture/replay initial changes to tt-metal for some workloads #17039

### What's changed

This is initial/bootstrapping changes (infra, tracing of ~16 Host APIs on host, device, unit tests e2e) for new feature "Light Metal capture/replay" that can trace host + device portion of workload and replay from standalone binary (tests do this, standalone runner exists too), will be built up on over next few months, and is to position ourselves favorably in automotive sector by having a way to pre-compile AI workload and run on-premise with no dependencies (one day).  Parent ticket for that feature for little bit more context is is: 

> [Feature Request] Light Metal Feature parent/tracking ticket #17037

For more detailed overview of changes (and caveat/restrictions) of changes here in this first PR, see below comment link and/or commit descriptions for slightly more detailed changes, or docs (.rst files) added for new APIs I added.

[More Details of Features/Changes here in this PR](https://github.com/tenstorrent/tt-metal/issues/17039#issuecomment-2610530734)

As mentioned in that link, I left several TODO(kmabee) items in code for things I want to fix next, this isn't comprehensive, and so I marked new APIs with disclaimer `Note: This LightMetal Trace feature is currently under active development and is not fully supported, use at own risk.`

### Notes / Explicit Callouts

#### Misc

@kmabeeTT  - I am going to update commit descriptions with issue number once approvals come and before merge, to reduce noise in ticket due to rebasing often. Will also sqush a couple more of the recent commits, left them unsquashed to be more visible for review.

@kmabeeTT - Been rebasing often (sometimes painful) past month, currently rebased on Jan22, so I am very recent.  Prefer to get this in and address any non-blocking concerns in follow up PR please.

#### use of detail::TraceDescriptor in host_api LoadTrace():

@ayerofieiev-tt / others - One of my new APIs used only at "replay time" is called `LoadTrace()` and takes a `detail::TraceDescriptor` object (to read it from flatbuffer binary and write it to device) which I understand is previously meant to be internal, Artem flagged this in open discussion on this PR already.  Not sure what proper fix is we like (please let me know) but nothing is broken right now with this, so would prefer to modify in follow up commit.   Couple options may include:

 - Move TraceDescriptor out of detail and into public API
 - Make it opaque by bouncing it through a void*, or some other API handle type to make it at least slightly more type safe

#### Public API header refactoring

I recently rebased on @afuller-TT change, and moved some of my hpp files under `tt_metal/api/tt-metalium/` but in honesty am not sure if I got it 100% right. I left a few inside tree, almost arbitrartily (wasn't sure how to decide whether they should be moved).  If nothing major, prefer to address afterwards.

```
-rw-r--r-- 1 kmabee  11K Jan 23 05:05 tt_metal/impl/flatbuffer/program_types_to_flatbuffer.hpp
-rw-r--r-- 1 kmabee 5.8K Jan 23 05:05 tt_metal/impl/flatbuffer/program_types_from_flatbuffer.hpp
-rw-r--r-- 1 kmabee 3.0K Jan 23 05:05 tt_metal/impl/flatbuffer/buffer_types_to_flatbuffer.hpp
-rw-r--r-- 1 kmabee 2.5K Jan 23 05:05 tt_metal/impl/flatbuffer/buffer_types_from_flatbuffer.hpp
-rw-r--r-- 1 kmabee 7.9K Jan 23 05:05 tt_metal/impl/flatbuffer/base_types_to_flatbuffer.hpp
-rw-r--r-- 1 kmabee 7.1K Jan 23 05:05 tt_metal/impl/flatbuffer/base_types_from_flatbuffer.hpp
```

#### Schema backward-compability

@omilyutin-tt - nothing done here for now, will continue to think about it in next steps.

### Checklist

No known fails.

- [x] Post commit CI passes (https://github.com/tenstorrent/tt-metal/actions/runs/12922657498)
- [x] Blackhole Post commit (https://github.com/tenstorrent/tt-metal/actions/runs/12922660098)
- [x] T3K Model Perf tests (https://github.com/tenstorrent/tt-metal/actions/runs/12922673184)
- [x] New/Existing tests provide coverage for changes

kmabeeTT committed:
LightMetal - Initial Cmake/Flatbuffers/Capture/Replay/Runner Infra for many popular APIs

 - This new "Light Metal Capture/Replay" feature is being built up still is aiming
   to provide infra to capture (host+device) workload to binary/disk and replay
   from binary (tt-metal for now, ttnn eventually).  The device side leverages
   "metal trace" feature and will serialize/deserialize/replay TraceDescriptors.

 - New Host APIs added to enable
   1. LightMetalBeginCapture() - start light metal binary capture
   2. LightMetalEndCapture() - end light metal binary capture and return blob to user
   3. LoadTrace() - load a TraceDescriptor to device (replay will use this API)

 - Instrument + support capture/replay few Host APIs: EnqueueTrace(), ReplayTrace(), LoadTrace()
 - Bulk of the work for this feature so far is Serialization/Deserialization
   of traced Host API calls to Flatbuffer binary.  Flatbuffer schema (.fbs files)
   defines the format, and ToFlatbuffer(), FromFlatbuffer() helper funcs convert.
 - Objects like Programs, Kernels, Buffers etc will be assigned unique
   global_id, and referred to by their global_id in capture + replay.
 - Capture code is in lightmetal_capture.cpp and replay is lightmetal_replay.hpp
 - standalone CLI runner lightmetal_runner to run binary from disk
 - Unrelated - Change trace_buffer.hpp to use fwd decl Buffer instead of buffer.hpp incl
   to reduce dependencies in lightmetal.cpp

 - When C++ define TT_ENABLE_LIGHT_METAL_TRACE!=1, host APIs not traced
   and also LightMetalBeginCapture() / LightMetalEndCapture() are NOPs

 - Add CMAKE_POSITION_INDEPENDENT_CODE=ON (-fPIC) for flatbuffer to avoid
   linking errors when lifting to tt-mlir project.
 - Use -Wno-restrict for flatbuffers compile to supress warning in g++12 build

 - PRCleanup: Remove light_metal_begin/end_capture() from device class, was not necessary
 - PRCleanup: Introduce LightMetalBinary type and use it instead of vector<uint8_t>
 - PRCleanup: Put fbs generated headers in flatbuffers subdir
 - PRCleanup: Make all FBS schema/structs to be tt::ttmetal::flatbuffer namespace

 LightMetal - Add initial Trace/Replay support for many popular host APIs

 - Not comprehensive, just initial changes for some of these paths:

   CreateBuffer()
   EnqueueWriteBuffer()
   EnqueueReadBuffer()
   Finish()
   DeallocateBuffer
   ReleaseTrace()
   CreateProgram()
   EnqueueProgram()
   CreateKernel()
   SetRuntimeArgs(uint32)
   SetRuntimeArgs(Kernel,RuntimeArgs)
   CreateCircularBuffer()

 - When Metal Trace is enabled, don't capture EnqueueProgram(), instead
   inject ReplayTrace(), would be used alongside LoadTrace()
 - Serialization / Deserialization of structs/enums defined in
   flatbuffer schema types.fbs handled via ToFlatbuffer() and
   FromFlatbuffer() functions.
 - Workaround in captureDeallocateBuffer() to ignore if CreateBuffer()
   not used (program binaries, trace buffer)
 - Fix clang-tidy build fails, use const shared_ptr<>& in Add*ToMap() functions

 - PRCleanup: Make all FBS schema/structs to be tt::ttmetal::flatbuffer namespace

snake_case to/from_flatbuffer functions

LightMetal - Add build_metal.sh --disable-light-metal-trace flag to set cmake/C++ define TT_ENABLE_LIGHT_METAL_TRACE=OFF/0

 - Keep "Light Metal Tracing" feature compile-time define set by default
   for simpler testing, but have quick option here to disable it if needed
   from perspective of host API "Capture" functions, and also the Light Metal
   enable/disable newly added APIS, they all become compile time NOP

kmabeeTT committed:
LightMetal - New test_lightmetal_sanity.cpp (add to CI) for LightMetalBinary capture+replay e2e testing

 - Metal trace tests (single op, two op) creates metal traces at capture time, runs at replay time
 - CI job passes for grayskull, wormhole, blackhole as expected
 - Use trace_buffer_size=4KB during capture/replay since BH needs it

kmabeeTT committed:
LightMetal - Add+Use TRACE_FUNCTION_ENTRY() macro to prevent recursive host API tracing

 - Add TRACE_FUNCTION_ENTRY() at the very start of function that is
   traced with TRACE_FUNCTION_CALL() to increment scope guard counter.

 - Really liked single macro usage per trace function, but some APIs
   like EnqueueProgram() and CreateDevice() (not currently traced, maybe
   one day) call other host APIs

 - Can't bundle with existing TRACE_FUNCTION_CALL() macro because
   sometimes it's called at end of traced function (when it needs to
   capture the return object) rather than beginning

ayerofieiev-tt commented on tt_metal/impl/device/device.hpp:
```diff
@@ -217,6 +219,11 @@ class Device {
     void set_trace_buffers_size(uint32_t size) { trace_buffers_size_ = size; }
 
+    // Light Metal
+    void light_metal_begin_capture();
+    std::vector<uint8_t> light_metal_end_capture();
+    void load_trace(const uint8_t cq_id, const uint32_t tid, detail::TraceDescriptor &trace_desc);
```
how are these traces different from traces on lines 214 - 218?

kmabeeTT replied:
Wasn't expecting others to see this quite yet while it's partially WIP, but the previous tracing APIs on lines 214-218 are the existing "metal trace" feature (been around since early 2024) which basically traces the device portion of the workload (ie. an on-device CQ).  The new stuff I'm adding for Light Metal is to trace the host part of the workload (ie. the host_api.hpp function calls to a binary), as well as leverage the on-device "metal trace" CQ part, so that (eventually) everything needed (host+device part of workload, precompiled binaries, etc) is in a single binary file to be rerun from disk.  My new APIs surround entire graph, wheras existing APIs surround only the EnqueueProgram() call, typically.

kmabeeTT replied:
Resolving this q from 2 weeks ago

kmabeeTT commented:
Forced push to address 4/8 bullets above.

nsmithtt commented on tt_metal/impl/lightmetal/host_api_capture_helpers.hpp:
```diff
@@ -0,0 +1,391 @@
+
+    // TODO (kmabee) - Workaround for now. Program Binaries buffer is created via Buffer::create() but can be
+    // deallocated on Program destruction while capturing is still enabled depending on test structure (scope).
+    // If we avoid recursively tracing host APIs this could go away.
```
I think you have resolved the recursive tracing problem, is this still an issue?

kmabeeTT replied:
Ah yeah, it is still an issue... comment isn't accurate, I will adjust it.  It's an issue because internal Buffers are created without traced API using internal `Buffer::create()` (so my "avoid recursive host API tracing" doesn't help here) but then deallocated via `DeallocateBuffer()` host API when goes is destroyed.   Not sure what the correct fix could be here... the only thing that comes to mind is them not caling DeallocateBuffer() , or me continuing to ignore them in this case (maybe more silently)?

kmabeeTT replied:
Talked to Nick offline, tend to agree, resolving.

nsmithtt commented on tt_metal/impl/lightmetal/lightmetal_replay.cpp:
```diff
@@ -0,0 +1,621 @@
+// TODO (kmabee) - Hardcode for now, eventually capture/replay "systemdesc" from binary.
+void LightMetalReplay::SetupDevices() {
+    log_debug(tt::LogMetalTrace, "LightMetalReplay(SetupDevices) - Using single chip WH device as temp hack.");
+    const size_t buffer_size = 2048;
```
Is there a default for this anywhere?

kmabeeTT replied:
There is a default actually, but it's zero (`DEFAULT_TRACE_REGION_SIZE=0`), so for now I need to override it with something non-zero. Going to add comment and update variable name.

nsmithtt reviewed (commented):
Hey Kyle this looks great to me! Few minor nits inline.

kmabeeTT commented on tt_metal/impl/buffers/circular_buffer_types.hpp:
```diff
@@ -18,12 +20,29 @@
+inline namespace v0 {
+class CircularBufferConfig;
+}
+inline flatbuffers::Offset<tt::target::CircularBufferConfig> ToFlatbuffer(
```
Hmm, I wasn't sure. I did do an update last week to my code where I tried to apply google style everywhere in my code thinking it would be the right thing to do (seeing a reference to it in a clang-format file, and discussions on slack) but yeah, it's not used everywhere in metal...

kmabeeTT replied:
Feedback from Bryan was metal has inconsistencies and ideally I follow advertised style in host_api.hpp (ie PascalCase, like I have today). Will leave this open till EoD for more feedback and close after that.

kmabeeTT commented on tt_metal/impl/lightmetal/binary.fbs:
```diff
@@ -0,0 +1,38 @@
+namespace tt.target.lightmetal;
+
+// Represents the Descriptor struct inside TraceDescriptor
+table Descriptor {
```
I agree, it's kind of vague, but does match the internally scoped struct inside tt-metal of the same name.  I'll update it to make it a bit less vague on our side, but leave the comment which describes what struct it corresponds to.

kmabeeTT commented on tt_metal/impl/lightmetal/host_api_capture_helpers.hpp:
```diff
@@ -0,0 +1,391 @@
+
+}  // namespace tt::tt_metal
+
+#ifdef ENABLE_TRACING
```
Thanks, that is fair will update -  Metal folks don't follow this anywhere, they have ifdef *EN* code all over their code.

kmabeeTT commented on tt_metal/impl/lightmetal/host_api_capture_helpers.hpp:
```diff
@@ -0,0 +1,391 @@
+#include "tracehost/types_to_flatbuffer.hpp"
+
+// FIXME (kmabee) - Temp hack, remove before merge and integrate as cmake define.
+#define ENABLE_TRACING 1
```
done now. 

Add build_metal.sh --enable-light-metal-trace flag to set cmake/C++ define TT_ENABLE_LIGHT_METAL_TRACE=ON/1

kmabeeTT commented:
I addressed all bullet points I planned on here for initial changes and cleaned up all the FIXMES last few days, and Nick's feedback, rebased on latest tt-metal that we're using in tt-mlir from today.

ayerofieiev-tt commented on tt_metal/distributed/mesh_device.hpp:
```diff
@@ -169,6 +169,11 @@ class MeshDevice : public IDevice, public std::enable_shared_from_this<MeshDevic
     void set_trace_buffers_size(uint32_t size) override;
 
+    // Light Metal
+    void light_metal_begin_capture() override;
```
Can you please consider to return a guard here instead that will end the trace in destructor unless it was ended explicitly by a call to guard?

kmabeeTT replied:
A RAII guard? I considered it after your comment, but it complicates things here especially when LightMetalBeginCapture() and LightMetalEndCapture() are called from different functions and would go out of scope (setup and teardown functions in a test fixture, or added to command-stream from tt-forge-fe at different points in time).  Currently my begin_capture() and end_capture() functions mimic how metal-trace begin/end APIs exist, would prefer them to stay ~aligned.

Side note - these APIs will move out of device class after suggestion from @cfjchu in meeting last week.

cfjchu replied:
great, thank you!

ayerofieiev-tt commented on tt_metal/distributed/mesh_device.hpp:
```diff
@@ -169,6 +169,11 @@ class MeshDevice : public IDevice, public std::enable_shared_from_this<MeshDevic
+    // Light Metal
+    void light_metal_begin_capture() override;
+    std::vector<uint8_t> light_metal_end_capture() override;
+    void load_trace(const uint8_t cq_id, const uint32_t tid, detail::TraceDescriptor& trace_desc) override;
```
What does `tid` argument stands for?

ayerofieiev-tt replied:
We are exposing something from `detail` namespace here

kmabeeTT replied:
Yep `tid` is trace_id, it matches same variable name used in all the other metal-trace related functions above this.  This function will be removed from device class, but it's called only by "light metal binary replay library" as host_api.hpp `LoadTrace()` function of same signature which has documentation, and tid is taken from the light-metal-binary, which was originally captured when the binary was created (ie. a trace_id and TraceDescriptor was captured at capture time, and then loaded to device at replay time here).

About `detail::TraceDescriptor`, yeah I don't love that, but I don't see another way to propagate the struct from the binary back into the device through an API.  I could move the `LoadTrace()` API out of host_api.hpp and into a `lightmetal_replay_utils.hpp` file (it feels like an organizational change, not sure if it helps).

kmabeeTT replied:
Done, made these changes and opened seperate [PR](https://github.com/tenstorrent/tt-metal/pull/17313) to merge that alone. Resolving this.

omilyutin-tt commented on tt_metal/impl/device/device.hpp:
```diff
@@ -179,6 +182,11 @@ class Device : public IDevice {
 
+    // Light Metal
+    void light_metal_begin_capture() override;
+    std::vector<uint8_t> light_metal_end_capture() override;
```
Alternatively, return the buffer object, and perform serialization at the caller side.

kmabeeTT replied:
This will move out of device class shortly, but even in it's new home, the idea was to keep things simple for all callers (C++, python) across tt-metal and tt-forge-fe that they would get just a vector of bytes.  Don't want to return a flatbuffer object either, and have flatbuffer (implementation details) leak out to caller.

omilyutin-tt replied:
How about:
```
struct LightTrace {
  std::vector<uint8_t> serialized_bytes;
};
```

kmabeeTT replied:
Sounds good, I created lightmetal_types.hpp, and moved it inside there as LightMetalBinary struct, updated all APIs (capture and replay) to use the new struct instead of std::vector<uint8_t> and moved some helper functions that lived elsewhere into the struct for reading from file and writing to file.

kmabeeTT committed:
LightMetal - New APIs LightMetalCompareToCapture() / LightMetalCompareToGolden() for verif

 - Put them in lightmetal_capture_utils.hpp since they are purely used
   at capture time, and not worthy enough to be inside host_api.hpp
   since just for verif.

 - Update test_lightmetal_sanity.cpp tests to use these API for
   functional correctness checking between capture + replay.

kmabeeTT commented on cmake/flatbuffers.cmake:
```diff
@@ -0,0 +1,16 @@
+# Function to generate FlatBuffers headers
+function(GENERATE_FBS_HEADER FBS_FILE)
+    get_filename_component(FBS_FILE_NAME ${FBS_FILE} NAME)
+    set(FBS_GENERATED_HEADER "${CMAKE_CURRENT_BINARY_DIR}/${FBS_FILE_NAME}_generated.h")
```
You're talking about the generated header files? They are currently under this:

```
./build_Debug/tt_metal/impl/command_generated.h
```

Will try adding a flatbuffers folder in there.

omilyutin-tt replied:
Yeah, this is mainly so that you limit the scope of CMake includes; a line like this is self explanatory:

```
target_include_directories(
    impl
    PUBLIC
        # Other stuff...
        ${CMAKE_CURRENT_BINARY_DIR}/flatbuffers
)
```

kmabeeTT replied:
Got it, thanks -  yes makes sense. Was able to make this change to limit scope alongside this change.

```
-        ${CMAKE_CURRENT_BINARY_DIR} # For generated flatbuffer headers
+        ${CMAKE_CURRENT_BINARY_DIR}/flatbuffers
```

kmabeeTT commented:
Rebased on latest main eef7b86207 from this morning (resolved bunch of public API reorg conflicts, moved some of my hpp files to public api folder), and removed light_metal_begin/end_capture() from device class, was not necessary there.

kmabeeTT commented on tt_metal/impl/tracehost/types.fbs:
```diff
@@ -0,0 +1,211 @@
+namespace tt.target;
+
+enum Arch: uint {
+  Grayskull = 0,
```
This is something briefly discussed with Nick before, tagging him @nsmithtt

nsmithtt replied:
@omilyutin-tt one pushback on this is that imo the introduction of a sentinel value in the API communicates that `Invalid` has some kind of meaning, but under no circumstances would `Invalid` be a legal programming for Arch name.  We should assert that arch == the runtime device arch which should handle all potential mis-programming of this value.

kmabeeTT replied:
I think this was just an example, but @omilyutin-tt  was referring to all enums.  However, in prior chats with @nsmithtt, that we ideally want to not support backward compatability, avoid it's complexities (should help with FuSa) and (not today, but eventually) just ensure capture + replay schema/changelistmatch.  Once we have that, could take it a step further, and ensure same flatbuffer and C++ so that we could one day use faster static_cast and avoid all the to_flatbuffer() and from_flatbuffer() conversion functions.

omilyutin-tt replied:
Yeah, it was just one example. The goal isn't solely for backwards compatibility, also to ensure integrity of the serialized values (distinguish "unset" fields). I found this guidance for protobufs that elaborates on the point https://protobuf.dev/best-practices/dos-donts/#unspecified-enum. However I think flatbuffers is a bit different here in that it will return a value that might not be known to the old client, instead of the default?

My perspective here is that we should treat these serialized buffers as untrusted, so we should be careful about making assumptions here. Especially since we are feeding them into runtime for execution:) 

Regarding backward compatibility. It may not be an issue for lightmetal, however the serialization primitives we are defining here will be more general. I imagine we will use these for communicating with distributed executors, for example, and I'm not sure we can ignore backward compatibility in that world. 

Having said that, we aren't at that stage, I agree that we should not overthink these problems. I assume the serialization format we are defining here can be broken, if necessary?

kmabeeTT replied:
Thansk @omilyutin-tt - You bring up some good points here, I think it's worthy of revisiting in seperate issue with discussion, but plan to skip it for purposes of this PR (initial bootstrap changes, to build upon, only used by me for short term).

kmabeeTT commented on tt_metal/impl/tracehost/command.fbs:
```diff
@@ -0,0 +1,117 @@
+// Define schema for tracing host API calls, called Commands in this context.
+include "tracehost/types.fbs";
+
+namespace tt.target;
```
Sure, I have made local change (global search and replace, update all usages, will force push soon, it's squashed into original commits) to do this:

```
-namespace tt.target;
+namespace tt.tt_metal.flatbuffer;
```

kmabeeTT commented on tt_metal/impl/tracehost/types.fbs:
Okay, this was a bit of a pain, but I have something that mostly matches original headers the enums were found in. Planning to use this breakdown for .fbs files (and then also breakdown the types_to_flatbuffer.hpp / types_from_flatbuffer.hpp into similarly named for each of these 5 groups):

```
// flatbuffer/buffer_types.fbs
// flatbuffer/core_types.fbs
// flatbuffer/base_types.fbs
// flatbuffer/kernel_types.fbs
// flatbuffer/runtime_types.fbs
```

kmabeeTT replied:
Do we feel strongly about splitting this file up?

Spent most of today on this item (quite a bit of work), almost done, but not quite (scared to squash it in case decision is later reverted)- the split from 3 files into 18 and the associated include spaghetti, is turning out to be brittle, and error prone since now there are dependencies between these fbs files and their associated to/from_flatbuffer.hpp files.  I have finally updated a bunch of header includes (a little bit of an explosion) to compile clean, but I know it's not 100% correct and it's only working because some headers are transitively included.  Fixing them will take some more work.

omilyutin-tt replied:
I appreciate you going through this effort, there is certainly a bit of work, I suspect partially due to the C++ primitives themselves not being clearly packaged.

The main goal here is to establish a forcing function to make future development easier. We want to avoid an encompassing "types" lib with everything in it, have a clear hierarchical structure with no spaghetti and circular dependencies... I think if we can scope out at least some of the logically cohesive functionality, that would be great. Perhaps, stuff for "buffers" and "core"? Also, I'd drop "types" from these names (ambiguous if "buffer types" refers to "different types of buffers" or "types related to buffer"), I think you can also group to + from into the same libs.

kmabeeTT replied:
Made changes, split to/from flatbuffer files the same way, going to resolve this.

kmabeeTT commented:
Open to approvals/reviews - Just rebased on latest main, my tests still passing locally and now the 8 successful PR checks are ticked,  I expect my previous passing checklist/CI runs from yesterday to be valid, no code changes on my side.

omilyutin-tt commented on tt_metal/api/tt-metalium/circular_buffer_types.hpp:
```diff
@@ -18,12 +20,30 @@
 using CBHandle = uintptr_t;
 
 class CircularBufferConfig {
+    friend flatbuffers::Offset<tt::tt_metal::flatbuffer::CircularBufferConfig> tt::tt_metal::ToFlatbuffer(
+        const tt::tt_metal::CircularBufferConfig& config, flatbuffers::FlatBufferBuilder& builder);
+    friend CircularBufferConfig FromFlatbuffer(
+        const tt::tt_metal::flatbuffer::CircularBufferConfig* config_fb, const Buffer* shadow_global_buffer);
```
Is there a reason we can't use public accessors for this? I'd prefer to have more of public functions than create a whole new friend-only path.

kmabeeTT replied:
I could have done that, but it felt like not the right solution.  This was thinking on the topic:

1. Not everything had public accessors (need everything in object)
2. it seemed wrong to add them just for this serialization/deserialization (don't want to give other users impression they are public-facing things)
3. using friend was in-line with how serialization libraries work from my understanding

Is this okay?

kmabeeTT replied:
Resolving. Opened another PR ([link](https://github.com/tenstorrent/tt-metal/pull/17353))

omilyutin-tt commented on tests/tt_metal/tt_metal/lightmetal/test_lightmetal_sanity.cpp:
```diff
@@ -0,0 +1,375 @@
+#include "lightmetal_capture_utils.hpp"
+
+using std::vector;
+using namespace tt;
+using namespace tt::tt_metal;
```
Let's instead encapsulate the tests in:
```
namespace tt::tt_metal {
namespace {
...
}
}
```

omilyutin-tt commented on tests/tt_metal/tt_metal/lightmetal/test_lightmetal_sanity.cpp:
```diff
@@ -0,0 +1,375 @@
+using namespace tt;
+using namespace tt::tt_metal;
+
+namespace lightmetal_test_helpers {
```
Can drop this, I think the functions here make it self-explanatory

omilyutin-tt commented on tests/tt_metal/tt_metal/lightmetal/test_lightmetal_sanity.cpp:
```diff
@@ -0,0 +1,375 @@
+
+}  // namespace lightmetal_test_helpers
+
+namespace lightmetal_basic_tests {
```
Instead of using namespaces, we should do something like

```
using LightMetalBasicTest = SingleDeviceLightMetalFixture;

TEST_F(LightMetalBasicTest, CreateBufferEnqueueWriteRead)`
```
... the same for the rest. What goes in `TEST_F` is registered with gmock, and you can conveniently use gfilters to run a subset of tests.

omilyutin-tt replied:
Also let's drop "Sanity" everywhere.

omilyutin-tt reviewed (changes requested):
There are many interesting changes here, and this is unfortunate that the PR is getting big. I appreciate the patience, especially given recent re-shuffling of public metal headers!

I think this PR needs to be split up, and the changes should go in incrementally. Can we scope out changes that cause the most conflicts during re-bases and have them in? I started reviewing the implementation - lots of new code and the corresponding comments, and I don't want that to cause more work for you in terms of re-basing and resolving conflicts.

As an example, changes to set up flatbuffers as a dependency in cmake files is a no-brainer and can go in quickly. Same for docs/... folder and the key APIs in `host_api.hpp`, as we already discussed them.

Changes to the metal implementation files likes device.cpp / tt_metal.cpp / etc can rely on empty stubs (e.g. `TRACE_FUNCTION_ENTRY` can be a no-op) for the time being, if we need to.

`lightmetal_runner.cpp` is another stand-alone change, as well as all files related to flatbuffers and serialization. Similarly, we can rely on an no-op, if these have to be used somewhere in the metal implementation where you want to avoid merge conflicts.

Does it make sense? I'd also really appreciate input from @ayerofieiev-tt @blozano-tt @afuller-TT on this. I love all of the prototyping work - this is critical to get the user stories right, but reviewing big PRs is non trivial and inherently slow, and chasing issues after the fact is error prone. One option is to leave this PR as is, and copy-over changes to the clean small PRs with targeted easy to review changes.

kmabeeTT commented on tt_metal/api/tt-metalium/lightmetal_types.hpp:
Do you mean the definition of the struct itself?  We'd discussed this offline previously but since I added helper functions to the struct and gave it a body (and it's used in a few places as output of capture and input to replay) it seems like defining a struct inside host_api.hpp (none are defined in there) wasn't the right idea anymore, no?

omilyutin-tt replied:
Right, apologies - you are right the functionality is getting too big now. Let's have it `light_metal_binary.hpp` though? The "types" doesn't force devs to think about where to put the new functionality, which may lead to chaos very quickly.

omilyutin-tt replied:
But let's chat on this kind of details on a dedicated review:)

kmabeeTT replied:
We had offline discussion, and I renamed it to lightmetal_binary.hpp and changed it to a class.

davorchap commented:
@kmabeeTT is this PR on the critical path for TT-Distributed?

Is yes, we should do a design review first, then do the PR review.

kmabeeTT commented:
> @kmabeeTT is this PR on the critical path for TT-Distributed?
> 
> Is yes, we should do a design review first, then do the PR review.

I didn't think so, adding @cfjchu to confirm. This Light Metal feature is orthogonal and is automotive inspired, but tt-distributed may leverage some of the flatbuffer serialization/deserialization work I have here, or at least build upon it.

aliuTT commented on tt_metal/impl/flatbuffer/base_types.fbs:
```diff
@@ -0,0 +1,102 @@
+namespace tt.tt_metal.flatbuffer;
+
+
```
We have definitions for a lot of these, for example in `tt_metal/api/tt-metalium/data_types.hpp`. Prefer to not duplicate types

kmabeeTT replied:
Yes, this is me duplicating all of them (and more in the other .fbs files) in flatbuffer schema format, which is unfortunately needed for anything that is to be serialized that it not plain old data type.  There is an idea, to refactor metal to remove the versions in `tt_metal/api/tt-metalium/data_types.hpp` you mention, and use the flatbuffer-generated .h file equivalents, but that would be a (large) follow up task.

aliuTT commented on tt_metal/api/tt-metalium/host_api.hpp:
```diff
@@ -883,6 +889,43 @@ void ReleaseTrace(IDevice* device, const uint32_t tid);
+// clang-format off
+/**
+ * Ends Light Metal binary capturing on host and all devices returns the binary blob to the user.
+ * Note: This LightMetal Trace feature is currently under active development and is not fully supported, use at own risk.
```
We have an experimental namespace for features under development `namespace experimental`, maybe better to make that explicit.

kmabeeTT replied:
I spawned this off to seperate [PR](https://github.com/tenstorrent/tt-metal/pull/17262) and merged it yesterday, droped experimental namespace since was colliding with others causing errors.

aliuTT commented on tt_metal/impl/CMakeLists.txt:
```diff
@@ -53,6 +70,7 @@ target_link_libraries(
     PUBLIC
         common
         Tracy::TracyClient
+        FlatBuffers::FlatBuffers
```
Is PUBLIC really needed here?

kmabeeTT replied:
I thought so, but now I'm not sure - I'll try changing it to PRIVATE.

kmabeeTT replied:
Yeah PRIVATE will make sense here, I have separate PR to merge this and related changes first, and it has this cleanup.

kmabeeTT commented:
>  think this PR needs to be split up, and the changes should go in incrementally

Thanks for feedback, @omilyutin-tt  will be looking into it, and answering some of your questions - It will be extra work, and my preference was to merge it all at once in one clean mege, but I will start to break this down into seperate PR's in some way that makes sense. It will reduce changes per PR, but may not have a huge impact to number of files, since majority of files are new, depending on how I split things up. I pulled out the flatbufers/cmake change and am opening a PR first for that.

aliuTT commented on tt_metal/api/tt-metalium/host_api.hpp:
```diff
@@ -883,6 +889,43 @@ void ReleaseTrace(IDevice* device, const uint32_t tid);
+ * | device       | The device to load the trace onto.                                     | IDevice *                     |                                    | Yes      |
+ * | cq_id        | The command queue id to load the trace onto.                           | uint8_t                       |                                    | Yes      |
+ * | trace_id     | A unique id to represent the trace on device.                          | uint32_t                      |                                    | Yes      |
+ * | trace_desc   | The trace descriptor to load onto the device.                          | detail::TraceDescriptor&      |                                    | Yes      |
```
We shouldn't have detail::TraceDescriptor in `host_api.hpp`, the intent was for this list of apis to be separate from the detailed apis. In general, I think all the light metal apis should be under detail namespace.

kmabeeTT replied:
Yeah, I knew about detail::TraceDescriptor problem and flagged it already for discussion.  One idea is to make it public.  Not sure about having light metal APIs in detrail namespace though, since light-metal capture/replay was to be built on top of tt-metal using public facing host_api.hpp only (so consumers like tt-forge-fe can enable/use it).  I think the detail APIs are only used internally, no?

aliuTT replied:
If light metal builds on top of tt-metal, it should be in a seperate project that depends on tt_metal, maybe like how TTNN structures it?

kmabeeTT replied:
We had agreement from a few folks in other discussons to move `TraceDescriptor` out of detail namespace so it can be used in this API.

About organization, it's a bit more nuanced than I said, some more details:

1. Light Metal capture / replay logic wil be (after I fix my code) private within tt-metal, but essentially trace the user-facing APs from host_api.hpp (from inside their implemementations)
2. Also, Light Metal APIs for initiating capture (and one for loading traces back to device) are public facing and built on top of tt-metal, so the capture and replay can be done by a user above tt-metal (like ttnn, or tt-forge-fe)

kmabeeTT replied:
Resolving here (this PR net changes is shrinking), I removed wrapping of TraceDescriptor with detail namespace and put up seperate PR ([link](https://github.com/tenstorrent/tt-metal/pull/17313)) for that along with this new API. Move discussion there if needed.

cfjchu commented:
> > @kmabeeTT is this PR on the critical path for TT-Distributed?
> > Is yes, we should do a design review first, then do the PR review.
> 
> I didn't think so, adding @cfjchu to confirm. This Light Metal feature is orthogonal and is automotive inspired, but tt-distributed may leverage some of the flatbuffer serialization/deserialization work I have here, or at least build upon it.

Agreed. We have been in the loop with @kmabeeTT during the bringup of this feature. We will piggy-back off overlapping components for this feature but we are not blocked on @kmabeeTT

aliuTT commented on tt_metal/impl/lightmetal/lightmetal_capture.cpp:
```diff
@@ -0,0 +1,236 @@
+
+#include <tt-metalium/logger.hpp>
+#include <tt-metalium/assert.hpp>
+#include "lightmetal_capture.hpp"
```
use the style of other includes `lightmetal/lightmetal_capture.hpp`

kmabeeTT replied:
This was done, after I refactored and moved to private cpp/hpp

omilyutin-tt commented:
We aren't blocked on this at the moment, but we will need the flatbuffers definitions for multi-host serialization at some point.

kmabeeTT commented on tt_metal/impl/flatbuffer/base_types_from_flatbuffer.hpp:
```diff
@@ -0,0 +1,140 @@
+#include "base_types_generated.h"
+
+namespace tt::tt_metal {
+inline namespace v0 {
```
Okay, should I update all of my code?  I had originally asked @blozano-tt on this very question 2 weeks ago and style from host_api.hpp was suggested (ie. PascalCase), so I updated everything to use it.

kmabeeTT replied:
Resolving, opened another PR with these updates (snake_case, inline v0 removed) [link](https://github.com/tenstorrent/tt-metal/pull/17353)

kmabeeTT commented on tt_metal/impl/flatbuffer/base_types_to_flatbuffer.hpp:
```diff
@@ -0,0 +1,149 @@
+        case BufferType::SYSTEM_MEMORY: return tt::tt_metal::flatbuffer::BufferType::SystemMemory;
+        case BufferType::L1_SMALL: return tt::tt_metal::flatbuffer::BufferType::L1Small;
+        case BufferType::TRACE: return tt::tt_metal::flatbuffer::BufferType::Trace;
+        default: throw std::invalid_argument("Unknown BufferType value in ToFlatbuffer()");
```
Sounds good, will make this change.

kmabeeTT replied:
Nice, confirmed this works. We see this compile error:

```
/localdev/kmabee/metal1/tt_metal/impl/flatbuffer/base_types_to_flatbuffer.cpp:11:13: error: enumeration value 'NEW_TYPE' not handled in switch [-Werror,-Wswitch]
   11 |     switch (type) {
      |             ^~~~
/localdev/kmabee/metal1/tt_metal/impl/flatbuffer/base_types_to_flatbuffer.cpp:18:1: error: non-void function does not return a value in all control paths [-Werror,-Wreturn-type]
   18 | }
      | ^
2 errors generated.
```

kmabeeTT replied:
Resolving - I opened another PR ([link](https://github.com/tenstorrent/tt-metal/pull/17353)) to merge flatbuffer schemas and to/flatbuffer and once merged will rebase this branch.

kmabeeTT commented on tt_metal/api/tt-metalium/host_api_capture_helpers.hpp:
```diff
@@ -0,0 +1,433 @@
+// when scope ends. Prevents recursive tracing of host APIs.
+struct TraceScope {
+    // Provide an inline definition in the header
+    static inline thread_local int depth = 0;
```
For thread safety; A singleton struct to track recursive host AP call depth, and in case host APIs happen to be called from multiple threads, we definitely want threads to have their own copy of the depth otherwise they'll step on each other. It's typical for this pattern from my understanding.

kmabeeTT commented on tt_metal/api/tt-metalium/lightmetal_capture_utils.hpp:
```diff
@@ -0,0 +1,72 @@
+
+namespace tt::tt_metal {
+
+// Note: LightMetalCompare functions could have been inside host_api.hpp / command_queue.cpp but seems better
```
Yeah, discussed with @nsmithtt today, and they are indeed intended to be public (and may be more APIs added here in the future), they are for users/application like the APIs in host_api.hpp.  The items in lightmetal_capture.hpp/cpp (soon) will be private.

kmabeeTT commented on tt_metal/api/tt-metalium/host_api_capture_helpers.hpp:
Thanks for this feedback, sounds good, will fix.

kmabeeTT commented on tt_metal/api/tt-metalium/lightmetal_capture.hpp:
```diff
@@ -0,0 +1,93 @@
+    LightMetalBinary CreateLightMetalBinary();
+    void Reset();
+
+    // Object Map Public Accessors
+    bool IsInMap(const Buffer* obj);
+    uint32_t AddToMap(const Buffer* obj);
+    void RemoveFromMap(const Buffer* obj);
+    uint32_t GetGlobalId(const Buffer* obj);
+    bool IsInMap(const Program* obj);
+    uint32_t AddToMap(const Program* obj);
+    void RemoveFromMap(const Program* obj);
+    uint32_t GetGlobalId(const Program* obj);
+    bool IsInMap(const Kernel* obj);
+    uint32_t AddToMap(const Kernel* obj);
+    void RemoveFromMap(const Kernel* obj);
+    uint32_t GetGlobalId(const Kernel* obj);
+    bool IsInMap(const CBHandle handle);
+    uint32_t AddToMap(const CBHandle handle);
+    void RemoveFromMap(const CBHandle handle);
+    uint32_t GetGlobalId(const CBHandle handle);
```
These are popular functions, used by users of `LightMetalCaptureContext` class, which is the helper functions in `host_api_capture_helpers.hpp` (which will soon be made private cpp/hpp).

I think lightmetal_capture.hpp (this file itself) also makes sense as private hpp. I moved it to the public API folder recently not knowing for sure, but now understanding more I will move it back alongside the cpp file.

kmabeeTT commented on tt_metal/api/tt-metalium/circular_buffer_types.hpp:
```diff
@@ -18,12 +20,30 @@
+class CircularBufferConfig;
+}
+namespace tt::tt_metal {
+inline namespace v0 {
+class CircularBufferConfig;
+}
+inline flatbuffers::Offset<tt::tt_metal::flatbuffer::CircularBufferConfig> ToFlatbuffer(
+    const tt::tt_metal::CircularBufferConfig& config, flatbuffers::FlatBufferBuilder& builder);
+inline CircularBufferConfig FromFlatBuffer(const tt::tt_metal::flatbuffer::CircularBufferConfig* config_fb);
```
It was intentional, to match the function definition in an hpp file.  But I learned now that it is redundant/not-required, so will remove it here for the forward declarations.

ayerofieiev-tt commented on tt_metal/impl/dispatch/command_queue.cpp:
```diff
@@ -589,6 +594,8 @@ void EnqueueWriteSubBuffer(
 
 void EnqueueProgram(
     CommandQueue& cq, Program& program, bool blocking) {
+    TRACE_FUNCTION_ENTRY();
+    TRACE_FUNCTION_CALL(CaptureEnqueueProgram, cq, program, blocking);
```
@kmabeeTT , after more review, the approach where we trace functions like looks wrong to me.
Honestly, instead, I would rather look into tracing CommandQueue itself. 
It looks to me, like all commands MUST go via Command Queue anyway.
More than that, there is already a big collection of operations defined in command_queue.hpp and what you desire is for them to be serialized.

Also what got me concerned during the review, some commands in command_queue have a flag "has_side_effect" and its set to true in some cases.

Lets discuss over a call on Tuesday, Jan 28.
Wdyt?

kmabeeTT replied:
Sure we can have a call, but in general this idea does not align with what we've been discussing since November while including @cfjchu, Paul and others where we said we wanted to trace (and replay) at highest level host_api.hpp functions only, to share with tt-distributed work.  The main point of the `TRACE_FUNCTION_CALL ()` "light metal" tracing is to capture the "host" part of the workload (host_api calls) , where existing "metal trace" feature captures CQ contents like you're talking about and is used to capture the "device part" of the workloads (which I then replay at replay-time using `EnqueueTrace()` to avoid calling `EnqueueProgram` the line you highlighted).  Further, many of the functions I trace there do not go through CQ at all, like `CreateProgram()`, `CreateBuffer()`, `CreateKernel()`, `CreateCircularBuffer()`, `SetRuntimeArgs()` to name a few.

FYI @nsmithtt

ayerofieiev-tt replied:
> Further, many of the functions I trace there do not go through CQ at all, like CreateProgram(), CreateBuffer(), CreateKernel(), CreateCircularBuffer(), SetRuntimeArgs() to name a few.

We are aligned here.

Please look specifically at changes in this file. Compare your tracing and `cq.run_command` calls. 
I don't see any difference.

ayerofieiev-tt replied:
I am removing CQ completely now. Ideally this part of traces have to happen inside HWCQ enqueue methods.

kmabeeTT replied:
Agreed `TRACE_FUNCTION_CALL()` macro and `cq.run_command()` are often side-by side, but the idea for consistency was to trace inputs to the highest level API functions (those in host_api.hpp) and replay them at that level for all APIs that "light metal trace/replay" supports, and not have a mixture of "some are traced/replay at host_api level, but some are traced/replay inside a HWCQ enqueue method", so that the executor library can solely use the public APIs in host_api.hpp to replay.
 
Currently, the underlying implementations of the host_api functions isn't too important since we're basically just capturing the inputs to these functions. Not seeing the problem yet.

Do you have preview available of your changes? Curious if `CommandQueue` is removed here, what the APIs will look like, here is `EnqueueProgram()` today as example:

```
void EnqueueProgram(
    CommandQueue& cq, Program& program, bool blocking) {
    TRACE_FUNCTION_ENTRY();
    TRACE_FUNCTION_CALL(CaptureEnqueueProgram, cq, program, blocking);
    detail::DispatchStateCheck(true);
    cq.run_command(
        CommandInterface{.type = EnqueueCommandType::ENQUEUE_PROGRAM, .blocking = blocking, .program = &program});
}
```

ayerofieiev-tt replied:
This is the draft https://github.com/tenstorrent/tt-metal/pull/17219/

I understand your direction with tracing high level api functions. 
I am sure this was aligned when metal api v1 was in progress. 
Everyone had an expectation that v1 will present metal c++ api this way, hiding details of implementation behind handles.
The effort was canceled. 
So realistically anyone can call to Buffer::create instead of CreateBuffer. 
This means that tracing in the CreateBuffer is guaranteed to miss creation of some buffers. 
Same for calls to CommandQueue.

Do you share this concern?

kmabeeTT replied:
Actually work here started and was indendent from V1 API knowledge (we found out about that later, so it didn't have much bearing).  

Yes though, you are right - it is a problem -  I raised the same point in our bi-weekly light-metal meeting with metal folks in Nov about the same thing (Buffer::create() instead of CreateBuffer() etc) asked whether I should be tracing Buffer::create() instead, and it was said that No, we only want to trace CreateBuffer() and force users to use only host_api.hpp public functions and not bypass them (an assert will fire in my feature if they do bypass it and use Buffer::create() or any other similar side stepping function, currently).

kmabeeTT replied:
We discussed offline and agreed to go forward with current changes to not be blocked here.

kmabeeTT committed:
Temp hack - make flatbuffers PUBLIC incdir until refactor soon (ignore)

kmabeeTT commented:
Forced push after rebasing on latest main (which included 2 small incremental PR's I merged shaved off from this big one), resolved some feedback, hope to resolve the rest today.

kmabeeTT commented on tt_metal/api/tt-metalium/lightmetal_types.hpp:
```diff
@@ -0,0 +1,64 @@
+
+    // Construct from raw data
+    explicit LightMetalBinary(std::vector<uint8_t> data) : data(std::move(data)) {}
+
```
Converted this to class with private data_ member, kept constructor.

kmabeeTT commented:
Closing this originally large PR - thanks all for feedback and offline discussions.  I split this up into 6 PR's that all depent on each other,  ([list](https://github.com/tenstorrent/tt-metal/issues/17039#issuecomment-2617393332)) incorporated feedback, and merged them over the 10 days.

kmabeeTT commented on tt_metal/api/tt-metalium/host_api_capture_helpers.hpp:
```diff
@@ -0,0 +1,433 @@
+// Debug Code                                               //
+//////////////////////////////////////////////////////////////
+
+inline void PrintHostDataType(const HostDataType& data) {
```
done. the split off smaller PR addressed this.
